third_party/move/move-vm/runtime/src/storage/struct_name_index_map.rs

```diff
@@ -1,43 +1,96 @@
+    /// if A == B, they have the same indices.
     pub(crate) fn struct_name_to_idx(&self, struct_name: StructIdentifier) -> StructNameIndex {
+        // Note that we take a write lock here once, instead of (*): taking a read lock, checking
+        // if the index is cached, re-acquiring the (write) lock, and checking again, as it makes
+        // things faster.
+        // Note that even if we do (*), we MUST check if another thread has cached the index before
+        // we reached this point for correctness. If we do not do this, we can end up evicting the
+        // previous index, and end up with multiple indices corresponding to the same struct. As
+        // indices are stored inside types, type comparison breaks!
         let mut index_map = self.0.write();
+
+        // Index is cached, return early.
         if let Some(idx) = index_map.forward_map.get(&struct_name) {
             return StructNameIndex(*idx);
         }
+
+        // Otherwise, the cache is locked and the struct name is not present. We simply add it
+        // to the cache and return the corresponding index.
         let idx = index_map.backward_map.len();
-        index_map.forward_map.insert(struct_name.clone(), idx);
-        index_map.backward_map.push(struct_name);
+        let prev_idx = index_map.forward_map.insert(struct_name.clone(), idx);
+        assert_none!(prev_idx, "Indexing map should never evict cached entries");
+
+        index_map.backward_map.push(Arc::new(struct_name));
         StructNameIndex(idx)
     }
 
-    pub(crate) fn idx_to_struct_name(&self, idx: StructNameIndex) -> StructIdentifier {
-        // TODO(loader_v2): Avoid cloning here, changed for now to avoid deadlocks.
+    /// Returns the reference of the struct name corresponding to the index. Here, we wrap the
+    /// name into an [Arc] to ensure that the lock is released.
+    pub(crate) fn idx_to_struct_name_ref(&self, idx: StructNameIndex) -> Arc<StructIdentifier> {
         self.0.read().backward_map[idx.0].clone()
     }
+
+    /// Returns the clone of the struct name corresponding to the index. The clone ensures that
+    /// the lock is released before the control returns to the caller.
+    pub(crate) fn idx_to_struct_name(&self, idx: StructNameIndex) -> StructIdentifier {
+        self.0.read().backward_map[idx.0].as_ref().clone()
+    }
+
+    /// Returns the struct tag corresponding to the struct name and the provided type arguments.
+    pub(crate) fn idx_to_struct_tag(
+        &self,
+        idx: StructNameIndex,
+        ty_args: Vec<TypeTag>,
+    ) -> StructTag {
+        let index_map = self.0.read();
+        let struct_name = index_map.backward_map[idx.0].as_ref();
+        StructTag {
+            address: *struct_name.module.address(),
+            module: struct_name.module.name().to_owned(),
+            name: struct_name.name.clone(),
+            type_args: ty_args,
+        }
+    }
+
+    #[cfg(test)]
+    fn len(self) -> usize {
```

Comment:
nit: get_checked_len or smt indicating what we do might be slightly better

Reply:
done